src/Categories/Tactic/Monoidal.agda

```diff
@@ -0,0 +1,444 @@
+    α′   : Expr ((A ⊗₀′ B) ⊗₀′ C) (A ⊗₀′ (B ⊗₀′ C))
+    α⁻¹′ : Expr (A ⊗₀′ (B ⊗₀′ C)) ((A ⊗₀′ B) ⊗₀′ C)
+    ƛ′   : Expr (unit′ ⊗₀′ A) A
+    ƛ⁻¹′ : Expr A (unit′ ⊗₀′ A)
+    ρ′   : Expr (A ⊗₀′ unit′) A
+    ρ⁻¹′ : Expr A (A ⊗₀′ unit′)
+
+  -- Embed a morphism in 'Expr' back into '𝒞' without normalizing.
+  [_↓] : Expr A B → (reify A) ⇒ (reify B)
+  [ id′ ↓]    = id
+  [ f ∘′ g ↓] = [ f ↓] ∘ [ g ↓]
+  [ f ⊗₁′ g ↓] = [ f ↓] ⊗₁ [ g ↓]
+  [ α′ ↓]     = associator.from
+  [ α⁻¹′ ↓]   = associator.to
+  [ ƛ′ ↓]     = unitorˡ.from
+  [ ƛ⁻¹′ ↓]   = unitorˡ.to
+  [ ρ′ ↓]     = unitorʳ.from
+  [ ρ⁻¹′ ↓]   = unitorʳ.to
+
+  infix 4 _≈↓_
+
+  -- TODO: is this sufficient or should we define an equality directly
+  -- on Expr?
+  _≈↓_ : (f g : Expr A B) → Set e
+  f ≈↓ g = [ f ↓] ≈ [ g ↓]
+
+  -- Invert a composition of coherence morphisms
+  invert : Expr A B → Expr B A
+  invert id′ = id′
+  invert (f ∘′ g) = invert g ∘′ invert f
+  invert (f ⊗₁′ g) = invert f ⊗₁′ invert g
+  invert α′ = α⁻¹′
+  invert α⁻¹′ = α′
+  invert ƛ′ = ƛ⁻¹′
+  invert ƛ⁻¹′ = ƛ′
+  invert ρ′ = ρ⁻¹′
+  invert ρ⁻¹′ = ρ′
+
+  -- Witness the isomorphism between 'f' and 'invert f'.
+  invert-isoˡ : ∀ (f : Expr A B) → [ invert f ↓] ∘ [ f ↓] ≈ id
+  invert-isoˡ id′ = identity²
+  invert-isoˡ (f ∘′ g) = begin
+    ([ invert g ↓] ∘ [ invert f ↓]) ∘ ([ f ↓] ∘ [ g ↓]) ≈⟨ cancelInner (invert-isoˡ f)  ⟩
+    [ invert g ↓] ∘ [ g ↓]                              ≈⟨ invert-isoˡ g ⟩
+    id                                                  ∎
+  invert-isoˡ (f ⊗₁′ g) = begin
+    ([ invert f ↓] ⊗₁ [ invert g ↓]) ∘ ([ f ↓] ⊗₁ [ g ↓]) ≈⟨ ⊗-elim (invert-isoˡ f) (invert-isoˡ g) ⟩
+    id                                                    ∎
+  invert-isoˡ α′   = associator.isoˡ
+  invert-isoˡ α⁻¹′ = associator.isoʳ
+  invert-isoˡ ƛ′   = unitorˡ.isoˡ
+  invert-isoˡ ƛ⁻¹′ = unitorˡ.isoʳ
+  invert-isoˡ ρ′   = unitorʳ.isoˡ
+  invert-isoˡ ρ⁻¹′ = unitorʳ.isoʳ
+
+  invert-isoʳ : ∀ (f : Expr A B) → [ f ↓] ∘ [ invert f ↓] ≈ id
+  invert-isoʳ id′ = identity²
+  invert-isoʳ (f ∘′ g) = begin
+    ([ f ↓] ∘ [ g ↓]) ∘ ([ invert g ↓] ∘ [ invert f ↓]) ≈⟨ cancelInner (invert-isoʳ g) ⟩
+    [ f ↓] ∘ [ invert f ↓]                              ≈⟨ invert-isoʳ f ⟩
+    id                                                  ∎
+  invert-isoʳ (f ⊗₁′ g) = begin
+    ([ f ↓] ⊗₁ [ g ↓]) ∘ ([ invert f ↓] ⊗₁ [ invert g ↓]) ≈⟨ ⊗-elim (invert-isoʳ f) (invert-isoʳ g) ⟩
+    id                                                    ∎
+  invert-isoʳ α′   = associator.isoʳ
+  invert-isoʳ α⁻¹′ = associator.isoˡ
+  invert-isoʳ ƛ′   = unitorˡ.isoʳ
+  invert-isoʳ ƛ⁻¹′ = unitorˡ.isoˡ
+  invert-isoʳ ρ′   = unitorʳ.isoʳ
+  invert-isoʳ ρ⁻¹′ = unitorʳ.isoˡ
+
+  invert-iso : ∀ (f : Expr A B) → Iso [ f ↓] [ invert f ↓]
+  invert-iso f = record
+    { isoˡ = invert-isoˡ f
+    ; isoʳ = invert-isoʳ f
+    }
+
+  NfWord : Set o
+  NfWord = List Obj
+
+  data NfExpr : NfWord → NfWord → Set o where
+    idⁿ : ∀ {N} → NfExpr N N
```

Comment:
I would be tempted, to avoid subst-juggling, to try
```agda
  idⁿ : ∀ {N M} → N == M → NfExpr N M
```
There's also some alternate way that @sstucki wrote to do this, but I can't recall where it is...

Reply:
@JacquesCarette I think you're thinking of `hid` which itself uses subst-juggling, so not sure it's helpful here.

But I was thinking: the use of subst in this module could maybe be avoided if the normal forms of the unitors and associators were defined by recursion on the domains following the same structure as the unit/associativity laws for lists themselves. That might also simplify the proofs of the corresponding "monoidality" laws which are proven by recursion and need helper lemmas to deal with the subst stuff. Anyway, just an idea, I don't know whether it would actually work out.

Reply:
Right, `hid`. If all `subst`-juggling can be isolated to one place, that counts as a win.